src/Java.Interop.Tools.JavaTypeSystem/JavaModels/JavaTypeName.cs

```diff
@@ -0,0 +1,151 @@
+using System;
+using System.Collections.Generic;
+using System.Linq;
+
+namespace Java.Interop.Tools.JavaTypeSystem.Models
+{
+	// An internal Type that represents a parsed Java type name structure.
+	// A type name string is passed to Parse() method to create this instance
+	// (the only place you can create an instance of this type), then it is
+	// used to create JavaTypeReference instance.
+	//
+	// The structure does not depend on any context type information.
+	//
+	// It is constructed from a "dotted" name.
```

Comment:
Do we *want* to be constructing things via "dotted" names and *not* via JNI names? Especially given the known ambiguities with dotted names?

Reply:
`JavaTypeName.cs` is scary, so it is c/p unchanged from `main` to ensure compatibility: https://github.com/xamarin/java.interop/blob/main/src/Xamarin.Android.Tools.ApiXmlAdjuster/JavaTypeName.cs.